Core/src/Propagator/detail/JacobianEngine.cpp

```diff
@@ -0,0 +1,228 @@
+         freeTransportJacobian * boundToFreeJacobian;
+}
+
+BoundMatrix boundToCurvilinearTransportJacobian(
+    const Vector3& direction, const BoundToFreeMatrix& boundToFreeJacobian,
+    const FreeMatrix& freeTransportJacobian,
+    const FreeVector& freeToPathDerivatives) {
+  // Calculate the derivative of path length at the the curvilinear surface
+  // w.r.t. free parameters
+  FreeToPathMatrix freeToPath = FreeToPathMatrix::Zero();
+  freeToPath.segment<3>(eFreePos0) = -1.0 * direction;
+  // Calculate the jacobian from global to local at the curvilinear surface
+  FreeToBoundMatrix freeToBoundJacobian = freeToCurvilinearJacobian(direction);
+  // Calculate the full jacobian from the local parameters at the start surface
+  // to curvilinear parameters
+  return freeToBoundJacobian *
+         (FreeMatrix::Identity() + freeToPathDerivatives * freeToPath) *
+         freeTransportJacobian * boundToFreeJacobian;
+}
+
+BoundToFreeMatrix boundToFreeTransportJacobian(
+    const BoundToFreeMatrix& boundToFreeJacobian,
+    const FreeMatrix& freeTransportJacobian) {
+  // Calculate the full jacobian, in this case simple a product of
+  // jacobian(bound->free) * jacobian(transport in free)
+  return (freeTransportJacobian * boundToFreeJacobian);
+}
+
+FreeToBoundMatrix freeToBoundTransportJacobian(
+    const GeometryContext& geoContext, const FreeVector& freeParameters,
+    const ActsMatrix<7, 8>& directionToAnglesJacobian,
+    const ActsMatrix<8, 7>& anglesToDirectionJacobian,
+    const FreeMatrix& freeTransportJacobian,
+    const FreeVector& freeToPathDerivatives, const Surface& surface) {
+  // Calculate the jacobian from free to bound at the final surface
+  FreeToBoundMatrix freeToBoundJacobian =
+      surface.freeToBoundJacobian(geoContext, freeParameters);
+  // Calculate the form factors for the derivatives
+  auto transport = freeTransportJacobian * anglesToDirectionJacobian;
+  FreeToPathMatrix sVec =
+      surface.freeToPathDerivative(geoContext, freeParameters);
+  // Return the jacobian to local
+  return freeToBoundJacobian *
+         (transport + freeToPathDerivatives * sVec * transport) *
+         directionToAnglesJacobian;
+}
+
+FreeToBoundMatrix freeToCurvilinearTransportJacobian(
+    const Vector3& direction, const ActsMatrix<7, 8>& directionToAnglesJacobian,
+    const ActsMatrix<8, 7>& anglesToDirectionJacobian,
+    const FreeMatrix& freeTransportJacobian,
+    const FreeVector& freeToPathDerivatives) {
+  // const ActsRowVector<3> normVec(direction);
+  const ActsMatrix<8, 7> transport =
+      freeTransportJacobian * anglesToDirectionJacobian;
+  auto sfactors =
+      direction.transpose() * transport.template topLeftCorner<3, 7>();
+
+  // Since the jacobian to local needs to calculated for the bound parameters
+  // here, it is convenient to do the same here
+  return freeToCurvilinearJacobian(direction) *
+         (transport - freeToPathDerivatives * sfactors) *
+         directionToAnglesJacobian;
+}
+
```

Comment:
Did I miss it or is `freeToFreeJacobian` missing?

Reply:
Yes, but isn't freeToFree just the application of the `freeTransportJacobian`? Hence, I thought having a dedicated method is a bit overkill, but somewhat we could still do it.

(i.e. if you wanted to optimise matrix multiplications you would only have to do so in one file)

Reply:
the freeToFree case is `freeTransportJacobian` multiplied with the 8x7 and 7x8 jacobians. As you have the direction derivatives \partial T_{final}/\partial T_{initial} (and also for the position) you have to consider the correlations.

Reply:
Oh, I see, will update it then.